### PR TITLE
[feature] Support create and sync newly added tables in cdc ingestion

### PR DESCRIPTION
**Pull request by qidian99**
### Purpose

This PR enables MySqlSyncDatabaseAction action to create and sync newly added tables as mentioned in https://github.com/apache/incubator-paimon/issues/1181.

### Tests
1. MySqlSyncDatabaseActionTest: test sync newly added tables

### API and Format
No.

### Documentation
No.

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
I think here we can have two modes for database sync.
1. One mode is static-tables. use `for (FileStoreTable table : tables) sink.sinkFrom(new DataStream<>(env, partitioned));` to create static sinks.
2. Another is `dynamic-tables` mode, static tables and newly tables all sink into new operators.

What do you think?

**Comment by qidian99** [comment]
Yes the idea of inducing a database sync mode is legit, and I refactored the code according. 

However, for dynamic-tables mode, we need to take more attention when routing static tables and newly tables into new operators, because 

1) the output tag of the new tables is determined at runtime, and
2) it's reasonable for users to switch back and forth between these two modes, e.g., using previous savepoint to start the database sync action.

Therefore, at this time being I route existing tables to single-table sinks and dynamic sink to multiple-tables sink.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java [hunk 1]
```diff
@@ -202,6 +202,9 @@ static MySqlSource<String> buildMySqlSource(Configuration mySqlConfig) {
         mySqlConfig
                 .getOptional(MySqlSourceOptions.HEARTBEAT_INTERVAL)
                 .ifPresent(sourceBuilder::heartbeatInterval);
+        mySqlConfig
+                .getOptional(MySqlSourceOptions.SCAN_NEWLY_ADDED_TABLE_ENABLED)
+                .ifPresent(sourceBuilder::scanNewlyAddedTableEnabled);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java:207]: Only set this when `MySqlDatabaseSyncMode` is DYNAMIC

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDebeziumJsonEventParser.java [hunk 2]
```diff
@@ -195,12 +217,86 @@ public List<DataField> parseSchemaChange() {
         return result;
     }
 
+    public Optional<Schema> parseNewSchema(String databaseName) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDebeziumJsonEventParser.java:220]: @Override

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/EventParser.java [hunk 3]
```diff
@@ -54,6 +60,16 @@ default String parseTableName() {
      */
     List<CdcRecord> parseRecords();
 
+    /**
+     * Parse newly added table schema from event.
+     *
+     * @param databaseName
+     * @return empty if there is no newly added table
+     */
+    default Optional<Schema> parseNewSchema(String databaseName) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/EventParser.java:69]: `parseNewTable`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java (hunk 1 above)
Comment (qidian99) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java:207]: > Only set this when `MySqlDatabaseSyncMode` is DYNAMIC

Done by setting the option in mysql config when MySqlDatabaseSyncMode is DYNAMIC

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java [hunk 4]
```diff
@@ -105,6 +110,10 @@ public class MySqlSyncDatabaseAction extends ActionBase {
     @Nullable private final Pattern includingPattern;
     @Nullable private final Pattern excludingPattern;
     private final Map<String, String> tableConfig;
+    private final String includingTables;
+    private final List<String> excludedTables;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java:114]: excludedTables can be local field in `build`, we can introduce `getMySqlSchemaList(List<String> excludedTables)`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java [hunk 5]
```diff
@@ -105,6 +110,10 @@ public class MySqlSyncDatabaseAction extends ActionBase {
     @Nullable private final Pattern includingPattern;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java:110]: So it should be non-null field? You should modify declarement and invoking place.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/EventParser.java [hunk 6]
```diff
@@ -54,6 +60,16 @@ default String parseTableName() {
      */
     List<CdcRecord> parseRecords();
 
+    /**
+     * Parse newly added table schema from event.
+     *
+     * @param databaseName
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/EventParser.java:66]: remove this line if you have no comment

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiTableParsingProcessFunction.java [hunk 7]
```diff
@@ -42,32 +50,103 @@
...
     }
 
     @Override
     public void processElement(T raw, Context context, Collector<Void> collector) throws Exception {
         parser.setRawEvent(raw);
+
+        // CDC Ingestion only supports single database at this time being.
+        //    In the future, there will be a mapping between source databases
+        //    and target paimon databases
+        String databaseName = parser.parseDatabaseName();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiTableParsingProcessFunction.java:108]: `parseDatabaseName` is no using?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java [hunk 8]
```diff
@@ -84,10 +99,54 @@ public void build() {
...
+                                        database, catalogLoader, tables, parserFactory, mode))
+                        .setParallelism(input.getParallelism());
+
+        // for newly-added tables, create a multiplexing operator that handles all their records
+        //     and writes to multiple tables
+        DataStream<CdcMultiplexRecord> newlyAddedTableStream =
+                SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed, CdcMultiTableParsingProcessFunction.DYNAMIC_OUTPUT_TAG);
+        // handles schema change for newly added tables
+        SingleOutputStreamOperator<Void> schemaChangeProcessFunction =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java:123]: schemaChangeProcessFunction no use

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiTableParsingProcessFunction.java [hunk 9]
```diff
@@ -42,32 +50,103 @@
 public class CdcMultiTableParsingProcessFunction<T> extends ProcessFunction<T, Void> {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiTableParsingProcessFunction.java:50]: Can you separate functions for static and dynamic mode?

**Comment by ChenShuai1981** [comment]
I think we need support paimon table name different with mysql table name, such as adding prefix like "ods_"  or suffix like "_ods" to indicate that those tables belong to the "operation data store" layer in data warehouse system, how do you think?

**Comment by ChenShuai1981** [comment]
I think we need to support the scenario that paimon table name different with mysql table name, such as adding prefix like "ods_"  or adding suffix like "_ods" to indicate that those paimon tables belong to the "operation data store" layer in data warehouse system, how do you think?

**Comment by qidian99** [comment]
> I think we need to support the scenario that paimon table name different with mysql table name, such as adding prefix like "ods_" or adding suffix like "_ods" to indicate that those paimon tables belong to the "operation data store" layer in data warehouse system, how do you think?

@ChenShuai1981 Yes, adding prefixes and suffixes is often needed in data ingestion. There are several other features that are missing at this time being:

1. add prefix and suffix to database and time
2. specify primary key or using unique key as primary key for tables w/o pk
3. merge multiple table from source into single table in Paimon
4. normalize source columns if contain invalid characters

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    public static final OutputTag<CdcMultiplexRecord> DYNAMIC_OUTPUT_TAG =
+            new OutputTag<>("paimon-dynamic-table", TypeInformation.of(CdcMultiplexRecord.class));
+    public static final OutputTag<Tuple2<Identifier, List<DataField>>>
+            DYNAMIC_SCHEMA_CHANGE_OUTPUT_TAG =
+                    new OutputTag<>(
+                            "paimon-dynamic-table-schema-change",
+                            TypeInformation.of(
+                                    new TypeHint<Tuple2<Identifier, List<DataField>>>() {}));
+
+    public CdcDynamicTableParsingProcessFunction(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java:73]: remove useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDatabaseSyncMode.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.paimon.flink.action.cdc.mysql;
+
+import java.io.Serializable;
+
+/**
+ * There are two modes for database sync. 1) STATIC mode, only write record from static tables.
+ * Newly added tables during runtime are not synced. 2) DYNAMIC mode, all records from static tables
+ * and newly added tables are routed into new operators.
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDatabaseSyncMode.java:26]: routed into new operators? new operator? just unified operator?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java [hunk 12]
```diff
@@ -85,55 +99,94 @@ public void build() {
...
+                        parsed, CdcDynamicTableParsingProcessFunction.DYNAMIC_OUTPUT_TAG);
+        // handles schema change for newly added tables
+        SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed,
+                        CdcDynamicTableParsingProcessFunction.DYNAMIC_SCHEMA_CHANGE_OUTPUT_TAG)
+                .process(new MultiTableUpdatedDataFieldsProcessFunction(catalogLoader));
+
+        FlinkStreamPartitioner<CdcMultiplexRecord> partitioner =
+                new FlinkStreamPartitioner<>(new CdcMultiplexRecordChannelComputer(catalogLoader));
+        PartitionTransformation<CdcMultiplexRecord> partitioned =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java:130]: no one use `partitioned`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java [hunk 13]
```diff
@@ -85,55 +99,94 @@ public void build() {
...
+        DataStream<CdcMultiplexRecord> newlyAddedTableStream =
+                SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed, CdcDynamicTableParsingProcessFunction.DYNAMIC_OUTPUT_TAG);
+        // handles schema change for newly added tables
+        SingleOutputStreamOperatorUtils.getSideOutput(
+                        parsed,
+                        CdcDynamicTableParsingProcessFunction.DYNAMIC_SCHEMA_CHANGE_OUTPUT_TAG)
+                .process(new MultiTableUpdatedDataFieldsProcessFunction(catalogLoader));
+
+        FlinkStreamPartitioner<CdcMultiplexRecord> partitioner =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java:128]: FlinkStreamPartitioner.partition

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiplexRecordChannelComputer.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    public int channel(CdcMultiplexRecord multiplexRecord) {
+        return Objects.hash(
+                        multiplexRecord.databaseName(),
+                        multiplexRecord.tableName(),
+                        computeChannelComputer(multiplexRecord).channel(multiplexRecord.record()))
+                % numChannels;
+    }
+
+    private ChannelComputer<CdcRecord> computeChannelComputer(CdcMultiplexRecord record) {
+        return channelComputers.computeIfAbsent(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcMultiplexRecordChannelComputer.java:60]: NPE here.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        return updatedDataFieldsOutputTags.computeIfAbsent(
+                tableName, CdcDynamicTableParsingProcessFunction::createUpdatedDataFieldsOutputTag);
+    }
+
+    public static OutputTag<List<DataField>> createUpdatedDataFieldsOutputTag(String tableName) {
+        return new OutputTag<>(
+                "new-data-field-list-" + tableName, new ListTypeInfo<>(DataField.class));
+    }
+
+    private OutputTag<CdcRecord> getRecordOutputTag(String tableName) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java:155]: remove lots of useless methods.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+ * function will capture newly added tables when syncing entire database and in cases where the
+ * newly added tables are including by attesting table filters.
+ *
+ * <p>This {@link ProcessFunction} can handle records for different tables at the same time.
+ *
+ * @param <T> CDC change event type
+ */
+public class CdcDynamicTableParsingProcessFunction<T> extends ProcessFunction<T, Void> {
+
+    private final EventParser.Factory<T> parserFactory;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcDynamicTableParsingProcessFunction.java:54]:
```
    public static final OutputTag<CdcMultiplexRecord> DYNAMIC_OUTPUT_TAG =
            new OutputTag<>("paimon-dynamic-table", TypeInformation.of(CdcMultiplexRecord.class));

    public static final OutputTag<Tuple2<Identifier, List<DataField>>>
            DYNAMIC_SCHEMA_CHANGE_OUTPUT_TAG =
            new OutputTag<>(
                    "paimon-dynamic-table-schema-change",
                    TypeInformation.of(
                            new TypeHint<Tuple2<Identifier, List<DataField>>>() {}));

    private final EventParser.Factory<T> parserFactory;
    private final String database;
    private final Catalog.Loader catalogLoader;

    private transient EventParser<T> parser;
    private transient Catalog catalog;
```

Is enough

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDatabaseSyncMode.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ */
+public enum MySqlDatabaseSyncMode implements Serializable {
+    STATIC,
+    DYNAMIC
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlDatabaseSyncMode.java:30]: Can we find better names to these two? Dynamic should be very confused with Dynamic bucket mode... (rebase master to see bucket mode in sink)

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * A {@link FlinkSink} which accepts {@link CdcRecord} and waits for a schema change if necessary.
+ */
+public class FlinkCdcMultiTableSink implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+    private static final String WRITER_NAME = "CDC MultiplexWriter";
+    private static final String GLOBAL_COMMITTER_NAME = "Multiplex Global Committer";
+    private final Lock.Factory lockFactory;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java:65]: static final  and final, should have a empty line.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    private final boolean isOverwrite = false;
+    private final Catalog.Loader catalogLoader;
+
+    public FlinkCdcMultiTableSink(Catalog.Loader catalogLoader, Lock.Factory lockFactory) {
+        this.catalogLoader = catalogLoader;
+        this.lockFactory = lockFactory;
+    }
+
+    private StoreSinkWrite.Provider createWriteProvider(CheckpointConfig checkpointConfig) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java:74]: checkpointConfig useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+                        + " to false.");
+        Preconditions.checkArgument(
+                env.getCheckpointConfig().getCheckpointingMode() == CheckpointingMode.EXACTLY_ONCE,
+                "Paimon sink currently only supports EXACTLY_ONCE checkpoint mode. Please set "
+                        + ExecutionCheckpointingOptions.CHECKPOINTING_MODE.key()
+                        + " to exactly-once");
+    }
+
+    protected OneInputStreamOperator<CdcMultiplexRecord, MultiTableCommittable> createWriteOperator(
+            StoreSinkWrite.Provider writeProvider, boolean isStreaming, String commitUser) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java:146]: isStreaming useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+        return new CdcRecordStoreMultiWriteOperator(
+                catalogLoader, writeProvider, commitUser, new Options());
+    }
+
+    // Table committers are dynamically created at runtime
+    protected SerializableFunction<
+                    String, Committer<MultiTableCommittable, WrappedManifestCommittable>>
+            createCommitterFactory(boolean streamingCheckpointEnabled) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcMultiTableSink.java:154]: streamingCheckpointEnabled useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/MultiTableUpdatedDataFieldsProcessFunction.java [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+                        SchemaChange.addColumn(
+                                newField.name(), newField.type(), newField.description(), null));
+            }
+        }
+        return result;
+    }
+
+    private void applySchemaChange(SchemaManager schemaManager, SchemaChange schemaChange)
+            throws Exception {
+        if (schemaChange instanceof SchemaChange.AddColumn) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/MultiTableUpdatedDataFieldsProcessFunction.java:144]: duplicate codes.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java (hunk 12 above)
Comment (qidian99) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java:130]: used down below:

```java
        if (parallelism != null) {
            partitioned.setParallelism(parallelism);
        }
```

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/FlinkCdcSyncDatabaseSinkBuilder.java:133]: I mean what does it do? I see it even has bugs in it, and I'm pretty sure no test can cover it.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java [hunk 23]
```diff
@@ -196,8 +196,8 @@ static MySqlSource<String> buildMySqlSource(Configuration mySqlConfig) {
                 .getOptional(MySqlSourceOptions.SERVER_TIME_ZONE)
                 .ifPresent(sourceBuilder::serverTimeZone);
         mySqlConfig
-                .getOptional(MySqlSourceOptions.SCAN_SNAPSHOT_FETCH_SIZE)
-                .ifPresent(sourceBuilder::fetchSize);
+                .getOptional(MySqlSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE)
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java:199]: why modify this?
add comments

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java [hunk 24]
```diff
@@ -210,6 +210,9 @@ static MySqlSource<String> buildMySqlSource(Configuration mySqlConfig) {
         mySqlConfig
                 .getOptional(MySqlSourceOptions.HEARTBEAT_INTERVAL)
                 .ifPresent(sourceBuilder::heartbeatInterval);
+        mySqlConfig
+                .getOptional(MySqlSourceOptions.SCAN_NEWLY_ADDED_TABLE_ENABLED)
+                .ifPresent(sourceBuilder::scanNewlyAddedTableEnabled);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlActionUtils.java:215]: why modify this?

Has `MySqlSourceBuilder.scanNewlyAddedTableEnabled`

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**